File: C/S14_BST/L2_PROBLEMS/14_inordersuccessor.c
#include <stdio.h>
#include <stdlib.h>
struct node
{
    int data;
    struct node *left;
    struct node *right;
};
struct node *createnode(int data)
{
    struct node *newnode = (struct node *)malloc(sizeof(struct node));
    newnode->data = data;
    newnode->left = NULL;
    newnode->right = NULL;
    return newnode;
}
struct node *min(struct node *root)
{
    if (root == NULL)
        return NULL;
    while (root && root->left)
    {
        root = root->left;
    }
    return root;
}
struct node *insert(struct node *root,int data)
{
    if(root==NULL)
    {
        root=createnode(data);
        return root;
    }
    if(root->data>data)
    {
        root->left=insert(root->left,data);
    }
    else if(root->data<data)
    {
        root->right=insert(root->right,data);
    }
    return root;
}
struct node *inordersuccessor(struct node *root, int data)
{
    if (root == NULL)
        return NULL;
    struct node *curr = root;
    struct node *succ = NULL;
    while (curr!=NULL)
    {
        if (curr->data > data)
        {
            succ = curr;
            curr = curr->left;
        }
        else if (curr->data < data)
        {
            curr = curr->right;
        }
        else
        {
            if (curr->right)
                return min(curr->right);
            else
                break;
        }
    }
    return succ;
}
int main()
{
    struct node *root=NULL;
    root=insert(root,67);
    root=insert(root,89);
    root=insert(root,809);
    root=insert(root,890);
    root=insert(root,829);
    struct node *s=inordersuccessor(root,89);
    if(s==NULL)printf("not exist");
    else
    printf("%d",s->data);
    return 0;

    
}

File: JAVA/S14_BST/L2_PROBLEMS/_14_inordersuccessor.java
class Node{
    int data;
    Node left;
    Node right;
    Node(int data){
        this.data=data;
        this.left=null;
        this.right=null;
    }
}
class Bst{
    Node insert(Node root,int data){
        Node newnode=new Node(data);
        if(root==null){
            return newnode;
        }
        if(root.data>data){
            root.left=insert(root.left,data);
        }
        if(root.data<data){
            root.right=insert(root.right,data);
        }
        return root;

    }
    public void inorder(Node root){
        if(root==null){
            return;
        }
        inorder(root.left);
        System.out.print(root.data+" ");
        inorder(root.right);
    }
    Node min(Node root){
        if(root==null)return null;
        Node curr=root;
        while(curr!=null && curr.left !=null){
            curr=curr.left;
        }
        return curr;
    }
    Node successor(Node root,int data){
        if(root==null)return null;
        Node curr=root;
        Node succ=null;
        while(curr!=null){
        if(curr.data>data){
            succ=curr;
            curr=curr.left;
            
        }
        else if(curr.data<data){
            curr=curr.right;
        }
        else{
            if(curr.right !=null){
                return min(curr.right);
            }
            else {
                break;
            }
        }
        }
        return succ;
    }
    
}
public class _14_inordersuccessor{
    public static void main(String[]args){
        Node root=null;
        Bst b=new Bst();
        root=b.insert(root,67);
        root=b.insert(root,6);
        root=b.insert(root,607);
        root=b.insert(root,60);
        root=b.insert(root,617);
        root=b.insert(root,63);
        Node s=b.successor(root,6);
        // b.inorder(root);
        System.out.print(s.data);
    }
}